File: frontend/services/product.ts
import apiClient from "@/services/api";
import { GlobalResponse } from "@/services/type/responses";

export type Product = {
    ID?: number;

    product_type: string;
    product_name: string;
    product_segment: string;

    warranty_length: number;
    stock?: number;
    price: number;

    specifications: string; // masih raw string dari backend

    headline: string;
    subheadline: string;
    product_code: string;

    // For fetched products this can be a base64 string; for uploads this will be a File
    display_image: File | string | undefined;
    ecommerce_url: string;
    shopee_url: string;
    tokopedia_url: string;
};

export type ProductType = 'laptop' | 'desktop' | 'server' | 'display' | 'tv' | 'iot' | 'aio' | ''

export type ProductSegment = "B2C" | 'B2G' | 'B2B';

export type NewProduct = {
    ID?: number;
    product_type: string;
    product_name: string;
    product_segment: ProductSegment;
    specifications: string;
    warranty_length: number;
    stock?: number;
    price: number;
    headline: string;
    subheadline: string;
    // Can be a File (when uploading) or string (when fetched from backend)
    display_image?: File | string;
    ecommerce_url: string;
    shopee_url: string;
    tokopedia_url: string;
}

export const addProduct = async (payload: Product) => {
    const formData = new FormData();

    Object.entries(payload).forEach(([key, value]) => {
        if (value !== undefined && value !== null) {
            formData.append(key, value instanceof File ? value : String(value));
        }
    });

    return apiClient.post(`/api/products/add`, formData, {
        headers: {
            "Content-Type": "multipart/form-data",
        },
    });
}

export const getProductByID = async (id?: number) => {
    return apiClient.get<GlobalResponse<Product>>(`/api/products/${id}`)
}

export type ProductFilter = {
    type: string;
    sort: 'price_asc' | 'price_desc' | 'name_asc' | 'name_desc' | '';
}

export const getProducts = async (params?: ProductFilter) => {
    return apiClient.get<GlobalResponse<Product[]>>(`/api/products/`, { params })
}

export const deleteProduct = async (productID: number) => {
    return apiClient.delete<GlobalResponse<null>>(`/api/products/${productID}`)
}

export const updateProduct = async (payload: Product) => {
    const formData = new FormData();

    Object.entries(payload).forEach(([key, value]) => {
        if (value !== undefined && value !== null) {
            formData.append(key, value instanceof File ? value : String(value));
        }
    });

    return apiClient.put<GlobalResponse<null>>(`/api/products/`, formData, {
        headers: {
            "Content-Type": "multipart/form-data",
        },
    });
}

File: backend-go/models/product.go
package models

import "gorm.io/gorm"

type ProductSegment string

const (
	B2B ProductSegment = "B2B"
	B2C ProductSegment = "B2C"
	B2G ProductSegment = "B2G"
)

type Product struct {
	gorm.Model
	Headline       string         `json:"headline"`
	Subheadline    string         `json:"subheadline"`
	ProductType    string         `gorm:"not null" json:"product_type"`
	ProductName    string         `gorm:"not null" json:"product_name"`
	ProductSegment ProductSegment `gorm:"not null" json:"product_segment"` //B2B, B2C
	WarrantyLength int            `gorm:"not null" json:"warranty_length"` // in days
	Stock          int            `gorm:"default:0" json:"stock"`
	Price          int            `gorm:"default:0" json:"price"`
	Specifications string         `gorm:"type:text" json:"specifications"`
	ProductCode    string         `json:"product_code"`
	DisplayImage   []byte         `gorm:"type:bytea" json:"display_image"`
	ECommerceURL   string         `json:"ecommerce_url"`
	ShopeeURL      string         `json:"shopee_url"`
	TokopediaURL   string         `json:"tokopedia_url"`
}
